### PR TITLE
Add a Coconet model

### PR DESCRIPTION
**Pull request by notwaldorf**
cc @jameswex as he's the original author.

### CODE REVIEW

**Comment by notwaldorf** [comment]
@adarob: i used the midi `instrument` to tell the 4 voices apart. do you think i should use `program` instead?

File: music/src/core/visualizer.ts [hunk 1]
```diff
@@ -141,11 +141,10 @@ export class Visualizer {
       const note = this.noteSequence.notes[i];
 
       // Size of this note.
-      const offset = this.config.noteSpacing * (i + 1);
-      const x = (this.getNoteStartTime(note) * this.config.pixelsPerTimeStep) +
-          offset;
+      const x = (this.getNoteStartTime(note) * this.config.pixelsPerTimeStep);
```
Comment (adarob) [inline music/src/core/visualizer.ts]: is this change intentional?

Comment (notwaldorf) [reply music/src/core/visualizer.ts]: it should be from a rebase

File: music/src/coconet/model.ts [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+  isInitialized() {
+    return this.initialized;
+  }
+
+  /**
+   * Sets up layer configuration from params
+   */
+  instantiateFromSpec() {
+    // Outermost dimensions' sizes of the non-final layers in the network.
+    const nonFinalLayerFilterOuterSizes = 3;
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: i think it's because they're always constant and don't depend on the spec changing. Should I just move them there?

File: music/src/coconet/model.ts [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+   */
+  async infill(sequence: INoteSequence) {
+    sequences.assertIsRelativeQuantizedSequence(sequence);
+
+    // Convert the sequence to a pianoroll.
+    const pianoroll = sequenceToPianoroll(sequence, DURATION_STEPS);
+
+    // Run sampling on the pianoroll.
+    const inputTensor = tf.tensor([pianoroll]) as tf.Tensor4D;
+    const samples = await this.run(inputTensor, this.spec.numSteps);
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: no, because tidy can't take an async function, and `this.run` needs to be awaited.

**Comment by notwaldorf** [comment]
🙋‍♀️ @nsthorat what are the best practices around `tidy()`? Should it only be called in the top level public methods, or basically at any level where you think you might leak tensors? Should I lean on over-tidying or no?

**Comment by notwaldorf** [comment]
@adarob this is ready for another review. I haven't done the tidy() changes yet, but I've addressed the other comments (I hope)

File: music/src/coconet/coconet_utils.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ * @param pianoroll Array of shape `[steps][NUM_PITCHES][NUM_VOICES]`,
+ * where each entry represents an instrument being played at a particular step
+ * and for a particular pitch. For example, `pianoroll[0][64] =[0, 0, 1, 0]`
+ * means that the third instrument plays pitch 64 at time 0.
+ * @returns A `NoteSequence` containing the melody.
+ */
+export function pianorollToSequence(pianoroll: tf.Tensor4D,
+    numberOfSteps: number): NoteSequence {
+  // First reshape the flat tensor so that it's shaped [steps][NUM_PITCHES][4].
+  const reshaped = tf.reshape(pianoroll,
```
Comment (adarob) [inline music/src/coconet/coconet_utils.ts]: this needs to be tidied or disposed

File: music/src/coconet/model.ts [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+
+  /**
+   * Replace a voice in a `NoteSequence` sequence with a different voice.
+   * @param sequence The `NoteSequence` to be changed.
+   * @param originalVoice The `NoteSequence` that will replace the notes in
+   * `sequence` with the same instrument.
+   * @return a new `NoteSequence` with sustained notes merged, and a voice
+   * replaced.
+   */
+  public replaceVoice(sequence: INoteSequence, originalVoice: INoteSequence)
```
Comment (adarob) [inline music/src/coconet/model.ts]: this should be in the utils, not here

File: music/src/coconet/model.ts [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+  /**
+   * Any consecutive notes of the same pitch are merged into a sustained note.
+   * Does not merge notes that connect on a measure boundary. This process
+   * also rearranges the order of the notes - notes are grouped by instrument,
+   * then ordered by timestamp.
+   *
+   * @param sequence The `NoteSequence` to be merged.
+   * @return a new `NoteSequence` with sustained notes merged.
+   */
+  public mergeHeldNotes(sequence: INoteSequence) {
```
Comment (adarob) [inline music/src/coconet/model.ts]: ditto on utils

File: music/src/coconet/model.ts [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+    let numStepsTensor: tf.Scalar = null;
+    if (!numSteps) {
+      numStepsTensor = this.numbersOfMaskedVariables(outerMasks).max();
+      numSteps = numStepsTensor.dataSync()[0];
+    } else {
+      numStepsTensor = tf.scalar(numSteps, 'float32');
+    }
+    let proll = pianorolls.clone();
+    for (let s = 0; s < numSteps; s++) {
+      const innerMasks = tf.tidy(() => {
```
[the remark this answers is not in the document]
Comment (adarob) [reply music/src/coconet/model.ts]: it still looks like we don't need this tidy since both of the inner functions tidy themselves...or we should remove the tidies from those functions (probably cleaner to do it that way)

File: music/src/coconet/model.ts [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+   * sequence.
+   * **Note**: regardless of the length of the notes in the original sequence,
+   * all the notes in the generated sequence will be 1 step long. If you want
+   * to clean up the sequence to consider consecutive notes for the same
+   * pitch and instruments as "held", as well as restore the original
+   * sequence, you can call `mergeHeldNotes()` on the result.
+   *
+   * @param sequence The sequence to continue. Must be quantized.
+   */
+  async infill(sequence: INoteSequence) {
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: note to future me: also add temperature.

**Comment by notwaldorf** [comment]
Wowowow thanks  @nsthorat for the super detailed explanation! This was great!

re: the vscode, i have `formatOnSave` on but it does crazy things for imports (it refuses to split them on more than one line), and more tragically, has completely stopped working in the last week. Given this public shaming, I'll try to fix it today 😅

**Comment by nsthorat** [comment]
Regarding `formatOnSave`, do you have this in your vscode settings: https://github.com/tensorflow/tfjs-core/blob/master/.vscode/settings.json#L29

Also you need to add clang-format as a dev-dependency: https://github.com/tensorflow/tfjs-core/blob/master/package.json#L23

And then you need to install 2 extensions in VSCode: clang-format, and TSLint.

If that doesn't work, lmk!

**Comment by notwaldorf** [comment]
I had all of those, turns out vscode was "grumpy" and i needed to restart my computer. Let's not talk about it.

File: music/src/coconet/coconet_utils.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * Converts a pianoroll representation to a `NoteSequence`. Note that since
+ * the pianoroll representation can't distinguish between multiple eighth notes
+ *  and held notes, the resulting `NoteSequence` won't either.
+ *
+ * @param pianoroll Array of shape `[steps][NUM_PITCHES][NUM_VOICES]`,
+ * where each entry represents an instrument being played at a particular step
+ * and for a particular pitch. For example, `pianoroll[0][64] =[0, 0, 1, 0]`
+ * means that the third instrument plays pitch 64 at time 0.
+ * @returns A `NoteSequence` containing the melody.
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/coconet_utils.ts]: ah, we mean different things by melody. will fix (here and errwhere)

File: music/src/coconet/coconet_utils.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  // First reshape the flat tensor so that it's shaped [steps][NUM_PITCHES][4].
+  const reshaped = tf.reshape(pianoroll,
+      [numberOfSteps, NUM_PITCHES, NUM_VOICES]);
+  const sequence = NoteSequence.create();
+  const notes: NoteSequence.Note[] = [];
+
+  for (let s = 0; s < numberOfSteps; s++) {
+    for (let p = 0; p < NUM_PITCHES; p++) {
+      for (let v = 0; v < NUM_VOICES; v++) {
+        const value = reshaped.get(s, p, v);
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/coconet_utils.ts]: i think the new `arraySync` in 0.15.2 should sync and use the correct shape, so i'll punt on this change until we merge https://github.com/tensorflow/magenta-js/pull/269

File: music/src/coconet/model.ts [hunk 11]
```diff
@@ -0,0 +1,589 @@
...
+    [1, 1], [2, 2], [4, 4], [8, 8], [16, 16], [16, 32],
+    [1, 1], [2, 2], [4, 4], [8, 8], [16, 16], [16, 32],
+    [1, 1], [2, 2], [4, 4], [8, 8], [16, 16], [16, 32]
+  ],
+  layers: null,
+  interleaveSplitEveryNLayers: 16,
+  numPointwiseSplits: 4,
+};
+
+class ConvNet {
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: Can't, it needs to be declared before where it's used :(

**Comment by nsthorat** [comment]
0.15.3 is released, let me know if you have any issues.

File: music/src/coconet/coconet_utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/coconet_utils.ts]: done!

**Comment by notwaldorf** [comment]
WHEW. This is ready for another pass. It even has 0 memory leaks! I quite like how the tidy's are done right now based on Nikhil's comments, so be gentle about that. @adarob 👀 again?

**Comment by notwaldorf** [comment]
The only thing I didn't change is mixing `await` and `then` in getting the checkpoint URL. i feel using only await is going to have more intermediate variables that aren't actually needed, but I don't really have a religious opinion about it otherwise. Should I change?

File: music/src/coconet/coconet_utils.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    const duration = note.quantizedEndStep - note.quantizedStartStep;
+    const voice = note.instrument;
+
+    if (voice < 0 || voice >= NUM_VOICES) {
+      logging.log(
+          `Found invalid voice ${voice}. Skipping.`, 'Coconet',
+          logging.Level.WARN);
+    } else {
+      for (let i = stepIndex; i < stepIndex + duration; i++) {
+        pianoroll[i][pitchIndex][voice] = 1;
```
Comment (adarob) [inline music/src/coconet/coconet_utils.ts]: throw error if i is >= numSteps

File: music/src/coconet/model.ts [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+          `NoteSequence ${sequence.id} does not have any notes to infill.`);
+    }
+    const numSteps = sequence.totalQuantizedSteps ||
+        sequence.notes[sequence.notes.length - 1].quantizedEndStep;
+
+    // Convert the sequence to a pianoroll.
+    const pianoroll = sequenceToPianoroll(sequence, numSteps);
+
+    // Run sampling on the pianoroll.
+    const samples = await this.run(pianoroll, this.spec.numSteps, temperature);
```
Comment (adarob) [inline music/src/coconet/model.ts]: numIterations

File: music/src/coconet/model.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: @nsthorat says the tidies should be in the inner functions, since they're the ones that allocate the memory. I'll remove this tidy.

[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/coconet/model.ts]: The way I understand it: `yaoSchedule` allocates some tensors so that it can calculate a result; this is why it needs to call `tidy`. It returns a tensor, which is why whoever is calling it needs to also call `tidy`.

**Comment by nsthorat** [comment]
One more high level thing: you should try to avoid `dataSync()` and use `data()` instead (or `array()`).

`dataSync()` blocks the UI thread while your operations execute, which can cause the rest of the page to jank out.

**Comment by notwaldorf** [comment]
@nsthorat donzo! 🎉

**Comment by notwaldorf** [comment]
closing for now.
